cli/pkg/helm/client.go

```diff
@@ -0,0 +1,566 @@
+package helm
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/spf13/pflag"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/downloader"
+	"helm.sh/helm/v3/pkg/getter"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/repo"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	apiextensionsV1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+var storage = repo.File{}
+
+// New returns a new Helm client with the provided options
+func New(options *Options) (Client, error) {
+	settings := cli.New()
+
+	err := setEnvSettings(options, settings)
+	if err != nil {
+		return nil, err
+	}
+
+	return newClient(options, settings.RESTClientGetter(), settings)
+}
+
+// newClient returns a new Helm client via the provided options and REST config
+func newClient(options *Options, clientGetter genericclioptions.RESTClientGetter, settings *cli.EnvSettings) (Client, error) {
+	err := setEnvSettings(options, settings)
+	if err != nil {
+		return nil, err
+	}
+
+	debugLog := func(_ string, _ ...interface{}) {}
+	if options.DebugLog != nil {
+		debugLog = options.DebugLog
+	}
+
+	actionConfig := new(action.Configuration)
+	err = actionConfig.Init(
+		clientGetter,
+		settings.Namespace(),
+		os.Getenv("HELM_DRIVER"),
+		debugLog,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &HelmClient{
+		Settings:     settings,
+		Providers:    getter.All(settings),
+		storage:      &storage,
+		ActionConfig: actionConfig,
+		linting:      options.Linting,
+	}, nil
+}
+
+// setEnvSettings sets the client's environment settings based on the provided client configuration
+func setEnvSettings(options *Options, settings *cli.EnvSettings) error {
+	if options == nil {
+		options = &Options{}
+	}
+
+	// set the namespace with this ugly workaround because cli.EnvSettings.namespace is private
+	// thank you helm!
+	if options.Namespace != "" {
+		pflags := pflag.NewFlagSet("", pflag.ContinueOnError)
+		settings.AddFlags(pflags)
+		err := pflags.Parse([]string{"-n", options.Namespace})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// AddOrUpdateChartRepo adds or updates the provided helm chart repository
+func (c *HelmClient) AddOrUpdateChartRepo(entry repo.Entry) error {
+	chartRepo, err := repo.NewChartRepository(&entry, c.Providers)
+	if err != nil {
+		return err
+	}
+
+	chartRepo.CachePath = c.Settings.RepositoryCache
+
+	_, err = chartRepo.DownloadIndexFile()
+	if err != nil {
+		return err
+	}
+
+	if c.storage.Has(entry.Name) {
+		log.Printf("WARNING: repository name %q already exists", entry.Name)
```

Comment:
This is just a warning log if helm repo exists and we try to add/update again we log it. But I am dropping this check as we trigger `AddUpdateCHart` because we intentionally want to add/update the existing repo.